rfc/state-file.md

```diff
@@ -0,0 +1,94 @@
+and only use the Constellation CLI for Kubernetes cluster management (Kubernetes versions, service versions, image upgrades etc.).
+
+## Format
+
+The state file will be in either YAML or JSON format, and contain a version number to keep track of necessary migrations or breaking formats.
+The file should be split into multiple sections for each relevant resource.
+
+Information that is only used to upgrade Terraform resources (e.g. IAM) should not be part of this file.
+
+The following is a list of proposed sections:
+
+* Cloud resources
+
+    Information about resources we currently create using Terraform.
+    Essentially, we want to record the information we currently write to the cluster ID file during `create`,
+    as well as the Terraform output during `create`.
+
+    This field is expected to be manually set if cloud resources were manually deployed by a user.
+    If cloud resource were created with the Constellation CLI (using `constellation create`),
+    it will be populated automatically.
+
+* Runtime resources
+
+    cluster ID, measurement salt etc.
+    This field will be automatically populated during `constellation init`
+
+YAML file example:
+
+```yaml
+version: "v1"
+cloudResources:
+    uid: "001122"
+    publicIP: "192.0.2.1"
+    initSecret: c2VjcmV0Cg==
+    apiServerCertSANs:
+        - "192.0.2.1"
+        - "endpoint.example.com"
+    azure:
+        resourceGroup: "example-group"
+        subscriptionID: "subscription-id"
+        networkSecurityGroupName: "example-group-name"
+        loadBalancerName: "example-loadbalancer-name"
+        userAssignedIdentity: "identity"
+        attestationURL: "attestation.example.com"
+    gcp:
+        ProjectID: "project-id"
+        IPCidrNode: "192.0.2.0/24"
+        IPCidrPod: "192.0.2.0/24"
+runtimeResources:
+    clusterID: "00112233445566778899AABBCCDDEEFF"
+    ownerID: "00112233445566778899AABBCCDDEEFF"
+    measurementSalt: "c2VjcmV0Cg=="
+```
+
+## Updates to the state file
+
+Values depending on Terraform, that change during a Constellation upgrade, should be automatically updated by the CLI if the cloud resources are not manually managed by the user.
+If they are manually managed, the user has to update the values on their own.
+
+Values depending purely on Kubernetes/node state should be automatically updated by the CLI.
+
+Breaking format changes should be automatically handled for `runtimeResources`.
+They should be automatically handled for `cloudResources` if the resources are managed by the Constellation CLI,
+otherwise we should provide documentation and/or tooling for migrating old file formats.
+
+## Migrating from terraform outputs + id file
+
+The Constellation version implementing the state file should provide an upgrade path for moving the Terraform outputs and cluster id file content into the new state file.
+This should happen at the beginning of `constellation upgrade apply`.
```

Comment:
I feel like we could make this a separate command from the beginning. That would be a step towards user-manager TF. And I feel like we have to implement the functionality anyways. Forces us to design the state migration code as a separated entity.

Reply:
A couple things about this:
First, we need to handle the id file separately anyways, which probably shouldn't happen in a CLI command dedicated to parsing terraform state.

I'm also not yet 100% sold on the idea of implementing a command to parse Terraform state files for the information needed for the Constellation state file.
While it could be a nice convenience for people who do use Terraform, ideally the state file would allow users to use any method for creating their cloud and IAM resources.
Adding a command specifically for Terraform would be a nice convenience, but IMO this shouldnt need to be a high priority.